submariner/multiple-cable-drivers.md

```diff
@@ -0,0 +1,192 @@
+}
+```
+
+The `ClusterConnectivityPolicy` CR defines the criteria for selecting which cluster pairs can connect. To resolve the
+connectivity policy for a remote cluster's `Endpoint`, Submariner would search for a policy whose left and right label
+selectors match the corresponding `Cluster` resources of the local and remote `Endpoints`. If any matching policy is found,
+the clusters are connected, otherwise they are not.
+
+If there are no policies defined, Submariner would default to full mesh connectivity.
+
+`ClusterConnectivityPolicy` resources would be created and maintained on the broker cluster and synced to each participating cluster.
+
+#### Examples
+
+- We only want certain clusters to access a database server in another cluster and no other clusters to connect to one another. We
+label the server and client `Clusters` appropriately and define a single policy with corresponding label selectors. Since there's
+no other policy to match client cluster pairs, they will not be connected.
+
+```yaml
+apiVersion: v1
+kind: ClusterConnectivityPolicy
+metadata:
+  name: client-server-policy
+spec:
+  leftClusterSelector:
+    database-role: server
+  rightClusterSelector:
+    database-role: client
+```
+
+- We want any cluster to access a database server in another cluster but no other clusters to connect. We label the server
+`Cluster` appropriately and define a single policy with the corresponding label selector and an empty selector to match all.
+
+```yaml
+apiVersion: v1
+kind: ClusterConnectivityPolicy
+metadata:
+  name: client-server-policy
+spec:
+  leftClusterSelector:
+    database-server: true
+  rightClusterSelector:
+```
+
+### Multiple cable drivers API
+
+This following CRD is proposed:
+
+```Go
+type ClusterConnectionPolicies struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec ClusterConnectionPoliciesSpec `json:"spec"`
+}
+
+type ClusterConnectionPoliciesSpec struct {
+    Policies []ClusterConnectionPolicy `json:"policies,omitempty"`
+}
+
+type ClusterConnectionPolicy struct {
+    // LeftClusterSelector identifies the Cluster resources representing the clusters on one end of a connection. An empty
+    // selector indicates wildcard, ie matches any cluster.
+    LeftClusterSelector metav1.LabelSelector `json:"leftClusterSelector,omitempty"`
+
+    // RightClusterSelector identifies the Cluster resources representing the clusters on the other end of a connection.
+    // An empty selector indicates wildcard, ie matches any cluster.
+    RightClusterSelector metav1.LabelSelector `json:"rightClusterSelector,omitempty"`
+
+    CableDriver *CableDriverSpec `json:"cableDriver,omitempty"`
+}
+
+type CableDriverSpec struct {
+    // Name of the cable driver implementation.
+    Name string `json:"name"`
+
+    // Options specifies key/value configuration parameters for the cable driver.
+    Options map[string]string `json:"options,omitempty"`
```

Comment:
Currently I am looking more at the mixing of cable drivers rather than the enablement + advertising of multiple cable drivers, but my thinking was to go down the route of using ConfigMaps... the reason being that every time I wanted to extend/modify the Custom Resource Definition I would need to release a new definition, as well as being able to override certain params if needed locally and being able to prioritise them base on the order they are loaded... 

it would be good to get alignment on best practice for Submariner as regards configuration parameters for the cable driver work (both multiple interface enablement as well as the mixing/combining of backends):

I was thinking that configuration parameters fall into two categories based on their scope:
Pod Local - these are parameters that hold configuration parameters specific to each Pod (values vary from Pod to Pod). 
Namespace Local - these are configuration parameters that you would like to share across a namespace (in this case the submariner-operator namespace). 

I was originally thinking that something like the following would be useful:
- To use environment variables/secrets injected into the Gateway Pod where a parameter will have a varying value in each Pod. (involves extending the submariner.io_submariners.yaml for some parameters)... 
 - To use a mounted ConfigMap/secrets for configurations that need to be shared across pods.
But it would be good to indicate if there is a preference for something else in terms of cable driver  configurations... 

I guess not all configuration parameters are created equally - so some rough guidelines would be useful...

Reply:
> Currently I am looking more at the mixing of cable drivers rather than the enablement + advertising of multiple cable drivers, but my thinking was to go down the route of using ConfigMaps... the reason being that every time I wanted to extend/modify the Custom Resource Definition I would need to release a new definition, 

Not necessarily - with a CRD you can easily make changes in a backwards compatible manner - we've done that several times. Regardless, you'd have the same backwards compatibility considerations using a `ConfigMap`. There's pros and cons to using `ConfigMap` vs CRD. A `ConfigMap` is suitable for simple key/value params but this is more complex than that I think.

> as well as being able to override certain params if needed locally and being able to prioritise them base on the order they are loaded...

How does a CRD preclude that?

Reply:
Yeah I appreciate what you are looking at with the policy structures is more complex than Key value pair params as such the CRD is more appropriate. I was specifically referring to using a ConfigMap for the map parameter for the Cable driver structure. Just trying to make sure we are aligned on that...